node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/Nodes.java

```diff
@@ -880,6 +889,166 @@ private Node requireNode(String hostname) {
         return node(hostname).orElseThrow(() -> new NoSuchNodeException("No node with hostname '" + hostname + "'"));
     }
 
+    /**
+     * Locks the children of the given node, the node itself, and finally takes the unallocated lock.
+     * <br>
+     * When taking multiple locks, it's crucial that we always take them in the same order, to avoid deadlocks.
+     * We want to take the most contended locks last, so that we don't block other operations for longer than necessary.
+     * This method does that, by first taking the locks for any children the given node may have, and then the node itself.
+     * (This is enforced by taking host locks after tenant node locks, in {@link #lockAndGetAll(Collection, Optional)}.)
+     * Finally, the allocation lock is taken, to ensure no new children are added while we hold this snapshot.
+     * Unfortunately, since that lock is taken last, we may detect new nodes after taking it, and then we have to retry.
+     * Closing the returned {@link RecursiveNodeMutexes} will release all the locks, and the locks should not be closed elsewhere.
+     */
+    public RecursiveNodeMutexes lockAndGetRecursively(String hostname, Optional<Duration> timeout) {
+        Set<Node> children = new HashSet<>(list().childrenOf(hostname).asList());
+        Optional<Node> node = node(hostname);
+
+        int attempts = 5; // We'll retry locking the whole list of children this many times, in case new children appear.
+        for (int attempt = 0; attempt < attempts; attempt++) {
+            NodeMutexes mutexes = null;
+            Mutex unallocatedLock = null;
+            try {
+                // First, we lock all the children, and the host; then we take the allocation lock to ensure our snapshot is valid.
+                List<Node> nodes = new ArrayList<>(children.size() + 1);
+                nodes.addAll(children);
+                node.ifPresent(nodes::add);
+                mutexes = lockAndGetAll(nodes, timeout);
+                unallocatedLock = timeout.map(db::lockInactive).orElseGet(db::lockInactive);
+                RecursiveNodeMutexes recursive = new RecursiveNodeMutexes(hostname, mutexes, unallocatedLock);
+                Set<Node> freshChildren = recursive.children.stream().map(NodeMutex::node).collect(Collectors.toSet());
+                Optional<Node> freshNode = recursive.parent.map(NodeMutex::node);
+                if (children.equals(freshChildren) && node.equals(freshNode)) {
+                    // No new nodes have appeared, and none will now, so we have a consistent snapshot.
```

Comment:
`lockAndGetAll(nodes, timeout)` will return a `mutexes` which is a subset of `nodes`.  The subset is possibly the same set as that denoted by `nodes`.  `recursive` will only return the children in `mutexes`.  Therefore `freshChildren` is a subset of `children`, and equal only if no children were removed.  This method doesn't handle the case where more children were added to the node before the unallocated lock were taken.

Reply:
Oops, great catch. Yes, I "optimised away a read" there 🤦

Reply:
Restoring to the original additional `list().childrenOf(hostname)`.